### ISSUE TITLE
FP16 GEMM vs CUDA: Performance Discrepancy on NVIDIA A100

### ISSUE THREAD

**Issue by hxu296** [YoungHyun197/ptq4vm#1]
Hello, thank you for the awesome work! I am trying to replicate the FP16 GEMM vs CUDA speedup. I noticed that in [cuda_sandbox.py](https://github.com/YoungHyun197/ptq4vm/blob/main/cuda_measure/cuda_sandbox.py), the performance is compared against FP32 GEMM by default. However, when I compare against FP16 GEMM by casting matrices A and B to fp16 before performing the PyTorch matmul, the CUDA GEMM experiences a slowdown compared to the FP16 GEMM. I am using an NVIDIA A100. Could you please guide me on what I might be doing wrong? Thank you!

Here is my updated test file as a reference
```
import os
import sys

import torch
import numpy as np
import time
import statistics
from tqdm import tqdm
import vim_GEMM

# os.environ['CUDA_LAUNCH_BLOCKING'] = "1"
os.environ["CUDA_VISIBLE_DEVICES"] = '0'

eps = 1e-8

b = 256
m = 197 # 197
k = 768 # 768 # 1536
n = 3072 # 3072 # 80

quantBits = 4 # 4 or 8
testTurn = 100

def int_quantizer_BWD(x, quantBits):
    '''
    original
    '''
    mx = x.abs().max()
    scale = mx / (2**(quantBits-1)-1)
    x = torch.clamp(x, -mx, mx) / (scale+eps)
    x_int = torch.round(x)
    return x_int

def main():

    FP_time = []
    kernel_time = []

    A = torch.Tensor(np.random.normal(1., 0.1, b*m*k)).reshape(b, m, k).to('cuda')
    B = torch.Tensor(np.random.normal(1., 0.1, n*k)).reshape(n, k).to('cuda')
    smooth_scale = torch.Tensor(np.random.normal(1., 0.1, k)).to('cuda')

    if quantBits == 4:
        B_int4 = torch.Tensor(np.random.normal(1., 0.1, n*k//2)).reshape(n, k//2).to('cuda')
        smooth_scale = torch.Tensor(np.random.normal(1., 0.1, k//2)).to('cuda')
    
    a_s = torch.Tensor(np.random.normal(1., 0.1, m)).reshape(m, 1).to('cuda')
    w_s = torch.Tensor(np.random.normal(1., 0.1, n)).reshape(n, 1).to('cuda')

    B_q = int_quantizer_BWD(B_int4, quantBits) if quantBits == 4 else int_quantizer_BWD(B, quantBits)

    for i in tqdm(range(testTurn)):
        A = A.to(torch.float16)
        B = B.to(torch.float16)
        start = time.time()
        gemm_fake = (A @ B.mT)
        torch.cuda.synchronize()
        FP_time.append(time.time() - start)
        
        A = A.to(torch.float32)
        B = B.to(torch.float32)
        B_q_int8 = B_q.type(torch.int8).contiguous()
        A = A.contiguous()
        start = time.time()
        gemm_real = vim_GEMM.vim_GEMM(A, \
                    B_q_int8, \
                    smooth_scale, \
                    a_s, \
                    w_s, \
                    16,
                    quantBits)
        torch.cuda.synchronize()
        kernel_time.append(time.time() - start)


    median_FP = statistics.median(FP_time)
    median_kernel = statistics.median(kernel_time)
    speedup = median_FP / median_kernel if median_kernel != 0 else float('inf')
    
    print(f"{median_FP:.8f} sec | FP GEMM")
    print(f"{median_kernel:.8f} sec | kernel")
    print(f"Speedup: {speedup:.2f}x")


if __name__ == '__main__':
    main()
```

**Comment by sungonuni** [comment]
Hello, thank you for your interest in our ptq4vm research. I'd like to provide two answers to your question.

First, for a fair speed comparison, matrices A and B should also be converted to torch.float16 before running vim_GEMM. The quantization overhead should be performed on torch.float16 as well to demonstrate the acceleration effect.

Second, the kernel may run slower for certain tensor sizes. This is based on the characteristics of cutlass - when the size of the MMA is not a common divisor of the tensor size, additional padding processes are required to match the sizes, causing the kernel to run slower. We already have a kernel implementation that resolves this issue, and we will upload it soon. Thank you.
